File: Code/Java/Cycle_Detection_In_Linked_List.java
/*    Floyd’s Cycle-Finding Algorithm 
      Approach: This is the fastest method and optimized approach to traverse a LinkedList:  
        * Traverse linked list using two pointers.
        * Move one pointer(slow_p) by one and another pointer(fast_p) by two.
        * If these pointers meet at the same node then there is a loop. If pointers do not meet then linked list doesn’t have a loop.   
*/
import java.io.*;
import java.util.*;
public class Cycle_Detection_In_Linked_List {
    // Creation of SinglyLinkedListNode 
    static class SinglyLinkedListNode {
        public int data;
        public SinglyLinkedListNode next;
        
        public SinglyLinkedListNode(int nodeData) {
            this.data = nodeData;
            this.next = null;
        }
    }
    // Creation of Singly Linked List with insertNode function  
    static class SinglyLinkedList {
        public SinglyLinkedListNode head;
        public SinglyLinkedListNode tail;

        public SinglyLinkedList() {
            this.head = null;
            this.tail = null;
        }
         
        public void insertNode(int nodeData) {
            SinglyLinkedListNode node = new SinglyLinkedListNode(nodeData);

            if (this.head == null) {
                this.head = node;
            } else {
                this.tail.next = node;
            }
            this.tail = node;
        }
    }
    static boolean hasCycle(SinglyLinkedListNode head) {
            if(head == null ){
                 return false;
              }
              SinglyLinkedListNode slow = head;
              SinglyLinkedListNode fast = head.next;
              while(slow != fast) {
                    if(fast == null || fast.next == null || slow == null ) {
                       return false;
                    }
                    slow = slow.next;
                    fast = fast.next.next;
              }
              return true;
    }
    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        System.out.println("Enter the test cases: ");
        int tests = scanner.nextInt();

        for (int testsItr = 0; testsItr < tests; testsItr++) {
        	System.out.println("Enter the pos ( is used to denote the index of the node that tail's next pointer is connected to ): ");
            int index = scanner.nextInt();
           
            SinglyLinkedList llist = new SinglyLinkedList();
            System.out.println("Enter the llist count: ");
            int llistCount = scanner.nextInt();

            for (int i = 0; i < llistCount; i++) {
            	System.out.println("Enter next node: ");
                int llistItem = scanner.nextInt();
                llist.insertNode(llistItem);
            }
          
          	SinglyLinkedListNode extra = new SinglyLinkedListNode(-1);
            SinglyLinkedListNode temp = llist.head;

            for (int i = 0; i < llistCount; i++) {
                if (i == index) {
                    extra = temp;
                }
                if (i != llistCount-1) {
                    temp = temp.next;
                }
            }
            temp.next = extra;

            boolean result = hasCycle(llist.head);
            if( result == true ) {
                System.out.println("The given LList has cycle. "); 
            }else {
            	System.out.println("The LList has no cycle "); 
            }
        }
        scanner.close();
    }
}
/*
Sample Input 1:
Enter the test cases: 
1
Enter the pos ( is used to denote the index of the node that tail's next pointer is connected to ): 
1
Enter the llist count: 
3
Enter next node: 
1
Enter next node: 
2
Enter next node: 
3
The given LList has cycle.

Sample Input 2:
Enter the test cases: 
1
Enter the pos ( is used to denote the index of the node that tail's next pointer is connected to ): 
-1
Enter the llist count: 
1
Enter next node: 
1
The LList has no cycle 

Complexity Analysis:  
    * Time complexity: O(N)
      Only one traversal of the loop is needed.
    * Auxiliary Space: O(1) 
      There is no space required.
*/

File: Code/Python/Cycle detection in linked list.py
#Program to Detect loop in Linked List;
#Algorithm is Floyd-Warshall Algorithm

#creating node class 
class Node:
    #creating constructor to create new node and assign data with new data and next with None
    def __init__(self,data):
        self.data=data
        self.next=None
        
#creating Linkedlist class for our main functions
class LinkedList:
    #constructor to assign new head of linked list
    def __init__(self):
        self.head=None
    
    #push function for adding new node in the list
    def push(self,new_data):
        new_node=Node(new_data)
        new_node.next=self.head
        self.head=new_node
    
    #printLinkedList for printing linked list if it not form loop
    def printLinkedList(self):
        temp=self.head
        while(temp is not None):
            print(temp.data)
            temp=temp.next
    
    #loop detection fucntion return true if detected 
    def Detectloop(self):
        slow_pointer=self.head
        fast_pointer=self.head
        while(slow_pointer and fast_pointer and fast_pointer.next):
            slow_pointer=slow_pointer.next
            fast_pointer=fast_pointer.next.next
            if(slow_pointer == fast_pointer):
                return True

#driver's Code
createLinkedList=LinkedList()
createLinkedList.push(2)
createLinkedList.push(3)
createLinkedList.push(5)
createLinkedList.push(1)

createLinkedList.head.next.next.next=createLinkedList.head
if(createLinkedList.Detectloop()==True):
    print("Loop detect")
else:
    print("Loop not detected")
    createLinkedList.printLinkedList()
    
    
#time Complexity:O(n)
#space complexity:O(1)

#test case 1:
# createLinkedList.push(12)
# createLinkedList.push(32)
# createLinkedList.push(5)
# createLinkedList.push(10)
# output:
# Loop not detected
# 10
# 5
# 32
# 12

# test case 2:
# createLinkedList.push(2)
# createLinkedList.push(3)
# createLinkedList.push(5)
# createLinkedList.push(1)

# createLinkedList.head.next.next.next=createLinkedList.head
# output:
#     Loop detect